File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        pair = {'(': ')', '{': '}', '[': ']'}
        
        for c in s:
            if c in pair:
                stack.append(c)
            else:
                if not stack or pair[stack[-1]] != c:
                    return False
                stack.pop()
        
        return len(stack) == 0
                    

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();
        HashMap<Character, Character> map = new HashMap<Character, Character>();
        map.put('(', ')');
        map.put('{', '}');
        map.put('[', ']');
        
        for (char curr : s.toCharArray()) {
            if (map.containsKey(curr)) {
                stack.push(curr);
            } else {
                if (stack.size() != 0 && curr == map.get(stack.peek())) {
                    stack.pop();
                } else {
                    return false;   
                }
            }
        }
        
        if (stack.size() == 0) return true;
        return false;
    }
}